conda_libmamba_solver/shards_subset.py

```diff
@@ -0,0 +1,126 @@
+
+    def neighbors(self, node: Node):
+        """
+        All neighbors for node.
+        """
+        discovered = set()
+        for shardlike in self.shardlikes:
+            if node.package in shardlike:
+                # check that we don't fetch the same shard twice...
+                shard = shardlike.fetch_shard(node.package)
+                for package in shard_mentioned_packages(shard):
+                    if package not in self.nodes:
+                        self.nodes[package] = Node(node.distance + 1, package)
+                        # by moving yield up here we try to only visit dependencies
+                        # that no other node already knows about. Doesn't make it faster.
+                        if package not in discovered:  # redundant with not in self.nodes?
+                            print(f"{json.dumps(node.package)} -> {json.dumps(package)};")
+                            yield self.nodes[package]
+                    if package not in discovered:
+                        pass
+                        # dot format valid ids: https://graphviz.org/doc/info/lang.html#ids (or quote string)
+
+                        # we might not require "in self.nodes" neighbors since
+                        # we don't need to find the shortest path
+
+                        # yield self.nodes[package]
+
+                    discovered.add(package)  # also doesn't make it faster
+
+    def outgoing(self, node: Node):
+        """
+        All nodes that can be reached by this node, plus cost.
+        """
+        for n in self.neighbors(node):
+            yield n, 1
+
+    def shortest(self, start_packages):
+        # nodes.visited and nodes.distance should be reset before calling
+        self.nodes = {package: Node(0, package) for package in start_packages}
+        unvisited = [(n.distance, n) for n in self.nodes.values()]
+        while unvisited:
+            original_priority, node = heapq.heappop(unvisited)
+            if (
+                original_priority != node.distance
+            ):  # pragma: no cover; didn't match what's in the heap
+                continue
+            if node.visited:  # pragma: no cover
+                continue
+            node.visited = True
+
+            for next, cost in self.outgoing(node):
+                if not next.visited:
+                    next.distance = min(node.distance + cost, next.distance)
+                    heapq.heappush(unvisited, (next.distance, next))
+
+
+def build_repodata_subset(tmp_path, root_packages, channels):
+    channel_data: dict[str, ShardLike] = {}
+    for channel in channels:
+        for channel_url in Channel(channel).urls(True, context.subdirs):
+            subdir_data = SubdirData(Channel(channel_url))
+            found = fetch_shards(subdir_data)
+            if not found:
+                repodata_json, _ = subdir_data.repo_fetch.fetch_latest_parsed()
+                found = ShardLike(repodata_json, channel_url)  # type: ignore
```

Comment:
I was reviewing https://github.com/conda/conda-libmamba-solver/pull/697, and I wondered why do we need a ShardLike interface? Then saw this code. Are we going to subset all repodata going forward? That will remove any benefits from the .solv cache in Unix systems that just memory map the file and introduce maybe unneeded overhead? Also in general, reading the full JSON just to subset it, and then write it back, only to be read again by libmamba... I guess it's faster to just pass it through.

So... what if we do _not_ use `ShardLike` for non-sharded channels and just keep the original repodata.json?

Reply:
I would rather parse them into mamba PackageRecord and skip json, that's possible right?
I have read blog posts saying memory mapping isn't that big a win on a lot of modern systems.
Using the original cached json path instead of a fresh subset-of-monolithic-repodata would be a simple choice when loading LibMambaIndexHelper; even though to subset the real sharded repodata we also need to compute the subset of fake sharded repodata.

Reply:
We would still need to initialize a bunch of Python objects to pass them to libmamba. My point is that if (in the case of non sharded repodata) we already have a JSON on disk ready to pass to the libmamba constructors, why bother doing all these extra steps.

Reply:
For sharded repodata, we could avoid the JSON writing overhead and construct the Python objects for libmamba straight from the raw dict. And then measure if that's faster. I guess there's a sweet number somewhere we can use as a threshold if the benefits are not clear in all cases.